### PR TITLE
secret channel test

### PR DESCRIPTION
**Pull request by slingamn**
Builds on @fiskfan1999's work in #134. (@fiskfan1999, sorry I stole this from you! It can be challenging to get these fully merge-ready, I thought I'd spare you the back-and-forth for now.)

### CODE REVIEW

File: irctest/server_tests/chmodes/secret.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+        "Likewise, secret channels are not listed
+        at all unless the client is a member of the channel in question."
+        """
+
+        def get_listed_channels(replies):
+            channels = set()
+            for reply in replies:
+                # skip pseudo-channel listed by ngircd:
+                if reply.command == RPL_LIST and reply.params[1] != "&SERVER":
```
Comment (FiskFan1999) [inline irctest/server_tests/chmodes/secret.py:19]: Thanks for handling this @slingamn . Maybe `reply.params[1][0] == "&"` will be able to sort out the other pesudo-channels from other daemons like test-irc2.

**Comment by FiskFan1999** [comment]
I noticed that during charybdis test, the assertion for when the non-chanop joined the secret channel failed in the following way:
```        self.joinChannel("second", "#gen")
        self.sendLine("second", "LIST")
        replies = self.getMessages("second")
        # Should be only one line with command RPL_LIST
>       self.assertEqual(get_listed_channels(replies), {"#gen"})
E       AssertionError: None
E       assert set() == {'#gen'}
E         Extra items in the right set:
E         '#gen'
E         Full diff:
E         - {'#gen'}
E         + set()```

I believe this means that this daemon has a bug in which when a non-chanop joins the secret channel, it doesn't show up when they run LIST even though it should according to RFC1459 et al. I just noticed the solanum test failed in the same way.

Do you think this is accurate?

File: irctest/server_tests/chmodes/secret.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+class SecretChannelTestCase(cases.BaseServerTestCase):
+    @cases.mark_specifications("RFC1459")
+    def testSecretChannelListCommand(self):
+        """
+        <https://datatracker.ietf.org/doc/html/rfc1459#section-4.2.6>
+
+        "Likewise, secret channels are not listed
+        at all unless the client is a member of the channel in question."
+        """
```
Comment (progval) [inline irctest/server_tests/chmodes/secret.py:13]: Please mark and quote Modern too.

https://modern.ircdocs.horse/#secret-channel-mode

File: irctest/server_tests/chmodes/secret.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+        def get_listed_channels(replies):
+            channels = set()
+            for reply in replies:
+                # skip pseudo-channels (&SERVER, &NOTICES) listed by ngircd
+                # and ircu:
+                if reply.command == RPL_LIST and reply.params[1].startswith("#"):
+                    channels.add(reply.params[1])
+            return channels
```
Comment (progval) [inline irctest/server_tests/chmodes/secret.py:22]: Could you share this function (or at least the filtering algo) with `server_tests/list.py`?

**Review by progval — CHANGES_REQUESTED** [review]

**Comment by progval** [comment]
Thanks to both of you!

**Pull request merged**